doc/site/guides/lua-language-server.markdown

```diff
@@ -0,0 +1,244 @@
+ * @param right number
+ * @param bottom number
+ * @return nil|number[] featureIDs
+ */
+int LuaUnsyncedRead::GetFeaturesInScreenRectangle(lua_State* L)
+{
+ // CPP code
+}
+```
+
+## IDE support
+
+The [recoil-lua-library](https://github.com/beyond-all-reason/recoil-lua-library) package provides type definitions for your editor.
+
+For working with Lua, we recommend using [Visual Studio Code](https://code.visualstudio.com/) with the [lua-language-server](https://marketplace.visualstudio.com/items?itemName=sumneko.lua) plugin.
+
+See the [Lua Language Server website](https://luals.github.io/#install) for other supported editors.
+
+## Extracting the library
+
+## The library
+
+The Lua library is a collection of type definitions that can be included in other projects (e.g. BAR). It is extracted from this codebase and written into [recoil-lua-library](https://github.com/beyond-all-reason/recoil-lua-library) repo.
+
+This library is also included in this repo as a [submodule](/recoil-lua-library/).
+
+You should not modify the submodule directly as it is auto-generated.
+
+## Doc comments within CPP files
+
+Special comments blocks are parsed by [lua-doc-extractor](https://github.com/rhys-vdw/lua-doc-extractor) and converted into [definition files](https://luals.github.io/wiki/definition-files/).
+
+Comment blocks must start with `/**`, and by convention each line starts with `*`.
+
+The body of the comment is translated literally, but some code generation is required. For these [custom tags](https://github.com/rhys-vdw/lua-doc-extractor?tab=readme-ov-file#custom-tags) must be used.
+
+Markdown is supported in all text.
+
+### Authored library files
+
+Hand-authored library files exist in `/rts/Lua/Library/`. These are appropriate to use when a Lua type is used in multiple CPP files.
+
+All files under `/rts/Lua/Library/` are directly copied into the library when the workflow runs.
+
+### GitHub Workflow generation.
+
+Exporting the library is automated by the [Generate Lua library workflow](/.github/workflows/generate-lua-library.yml). You can test it locally by following the same steps locally.
+
+See doc [README](/doc/site/README.md) for info on doc generation.
+
+### Manually generating the library
+
+To test extracting library files from CPP, you can run [lua-doc-extractor](https://github.com/rhys-vdw/lua-doc-extractor) locally.
+
+First install `lua-doc-extractor`:
+
+```bash
+npm install -g lua-doc-extractor
+```
+
+At root, run:
+
+```bash
+# First delete any previously generated output.
+rm -rf rts/Lua/library/generated
+
+# Regenerate the library.
+lua-doc-extractor rts/Lua/*.cpp --dest rts/Lua/library/generated
+```
+
+## Documenting the codebase
+
+[Full list of annotations](https://luals.github.io/wiki/annotations/).
+
+### Common annotations
+
+- [`@function`](https://github.com/rhys-vdw/lua-doc-extractor?tab=readme-ov-file#function-name)
+- [`@table`](https://github.com/rhys-vdw/lua-doc-extractor?tab=readme-ov-file#table-name)
```

Comment:
Would be good if this didn't link to a personal repo.

Reply:
Agreed. I'm open to moving it to into the BAR org in future. If I do the link will redirect. I haven't done it yet because:

1. There is no BAR npm org (and I didn't feel I was the right person to take that initiative).
2. There's a bit of involvement getting admin rights and such.
3. BAR org disallows pushing with my git GUI and I don't want to introduce more friction when I'm trying to move fast.
4. I want to fix the remaining issues.
5. It's not blocking this.

Reply:
As per your other suggestion I can just remove this reference entirely since it's an internal tool.

Reply:
Honestly, idk why BAR is discussed as an org suitable for engine concerns, but as long as shit works I guess

Reply:
Sorry you're right. It should go into a recoil org.

Reply:
But for now the source of truth is my personal tool for extracting docs from recoil. I don't think there's any action here for this PR.